### PR TITLE
[MXNET-984] Add Java NDArray and introduce Java Operator Builder class

### PR DESCRIPTION
**Pull request by lanking520**
## Description ##
Just a copy of https://github.com/apache/incubator-mxnet/pull/12772
@andrewfayres @zachgk @piyushghai @nswamy @yzhliu 
### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### CODE REVIEW

File: scala-package/core/src/test/java/org/apache/mxnet/javaapi/NDArrayTest.java [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+    @Test
+    public void testCreateNDArray() {
+        NDArray nd = new NDArray(new float[]{1.0f, 2.0f, 3.0f},
+                new Shape(new int[]{1, 3}),
+                new Context("cpu", 0));
+        int[] arr = new int[]{1, 3};
+        assertTrue(Arrays.equals(nd.shape().toArray(), arr));
+        assertTrue(nd.at(0).at(0).toArray()[0] == 1.0f);
+        List<Float> list = new ArrayList<Float>();
+        list.add(1.0f);
```
Comment (nswamy) [inline scala-package/core/src/test/java/org/apache/mxnet/javaapi/NDArrayTest.java]: you can use Arrays.asList to create List with values in 1 line
`List<Float> list = Arrays.asList(1.0f,2.0f,3.0f);

File: scala-package/core/src/test/java/org/apache/mxnet/javaapi/NDArrayTest.java [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+                new Shape(new int[]{1, 3}),
+                new Context("cpu", 0));
+        assertTrue(Arrays.equals(nd.shape().toArray(), arr));
+    }
+
+    @Test
+    public void testZeroOneEmpty(){
+        NDArray ones = NDArray.ones(new Context("cpu", 0), new int[]{100, 100});
+        NDArray zeros = NDArray.zeros(new Context("cpu", 0), new int[]{100, 100});
+        NDArray empty = NDArray.zeros(new Context("cpu", 0), new int[]{100, 100});
```
Comment (nswamy) [inline scala-package/core/src/test/java/org/apache/mxnet/javaapi/NDArrayTest.java]: did you mean `NDArray empty = NDArray.empty..`?

File: scala-package/macros/src/main/scala/org/apache/mxnet/javaapi/JavaNDArrayMacro.scala [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+
+import org.apache.mxnet.init.Base._
+import org.apache.mxnet.utils.CToScalaUtils
+
+import scala.annotation.StaticAnnotation
+import scala.collection.mutable.ListBuffer
+import scala.language.experimental.macros
+import scala.reflect.macros.blackbox
+
+private[mxnet] class AddJNDArrayAPIs(isContrib: Boolean) extends StaticAnnotation {
```
Comment (nswamy) [inline scala-package/macros/src/main/scala/org/apache/mxnet/javaapi/JavaNDArrayMacro.scala]: Can you paste the generated NDArrayAPI somewhere(may be a gist)?

File: scala-package/macros/src/main/scala/org/apache/mxnet/javaapi/JavaNDArrayMacro.scala [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+      var OptionArgDef = ListBuffer[String]()
+      // Construct function Implementation field (e.g norm)
+      var impl = ListBuffer[String]()
+      impl += "val map = scala.collection.mutable.Map[String, Any]()"
+      // scalastyle:off
+      impl += "val args= scala.collection.mutable.ArrayBuffer.empty[org.apache.mxnet.NDArray]"
+      // scalastyle:on
+      // Construct Class Implementation (e.g normBuilder)
+      var classImpl = ListBuffer[String]()
+      ndarrayfunction.listOfArgs.foreach({ ndarrayarg =>
```
Comment (nswamy) [inline scala-package/macros/src/main/scala/org/apache/mxnet/javaapi/JavaNDArrayMacro.scala]: nit: `ndarrayarg`->`ndarrayArg`

File: scala-package/macros/src/main/scala/org/apache/mxnet/utils/CToScalaUtils.scala [hunk 5]
```diff
@@ -21,19 +21,19 @@ private[mxnet] object CToScalaUtils {
...
-      case "Shape(tuple)" | "ShapeorNone" => "org.apache.mxnet.Shape"
+      case "Shape(tuple)" | "ShapeorNone" => s"org.apache.mxnet.$shapeType"
       case "Symbol" | "NDArray" | "NDArray-or-Symbol" => returnType
       case "Symbol[]" | "NDArray[]" | "NDArray-or-Symbol[]" | "SymbolorSymbol[]"
       => s"Array[$returnType]"
-      case "float" | "real_t" | "floatorNone" => "org.apache.mxnet.Base.MXFloat"
-      case "int" | "intorNone" | "int(non-negative)" => "Int"
-      case "long" | "long(non-negative)" => "Long"
-      case "double" | "doubleorNone" => "Double"
+      case "float" | "real_t" | "floatorNone" => "java.lang.Float"
```
Comment (nswamy) [inline scala-package/macros/src/main/scala/org/apache/mxnet/utils/CToScalaUtils.scala]: Scala Float is not the same as java.lang.Float, have you tested compatibility between the two i.e., no precision loss?

Comment (lanking520) [reply scala-package/macros/src/main/scala/org/apache/mxnet/utils/CToScalaUtils.scala]: https://www.scala-lang.org/api/2.12.3/scala/Float.html and there should be no loss

File: scala-package/macros/src/main/scala/org/apache/mxnet/javaapi/JavaNDArrayMacro.scala (hunk 3 above)
Comment (lanking520) [reply scala-package/macros/src/main/scala/org/apache/mxnet/javaapi/JavaNDArrayMacro.scala]: In my next PR, I will add documentation generation

Comment (lanking520) [reply scala-package/macros/src/main/scala/org/apache/mxnet/javaapi/JavaNDArrayMacro.scala]: Here is the gist to that https://gist.github.com/lanking520/f9ed0eeaa45f2364471f1f3c6c9d598f

**Comment by nswamy** [comment]
reluctantly agreeing to adding Builders for all APIs with the premise that this design will be revisited before it makes to master.

**Comment by nswamy** [comment]
also please add more unit tests for these builder based APIs

**Pull request merged**